src/engine/platform/mixer.cpp

```diff
@@ -47,6 +55,42 @@ struct AudioStruct {
     SDL_AudioDeviceID dev;
 } audio_struct = {};
 
+Effect create_delay(f32 feedback, f32 delay_time) {
+    ASSERT(delay_time < CHANNEL_BUFFER_LENGTH_SECONDS, "Delay time is longer than channel buffer");
+    auto delay_func = [] (Effect *effect, f32 *buffer, u32 start, u32 len) -> void {
+        if (effect->delay._delay_len_seconds != effect->delay._prev_delay_len_seconds) {
+            effect->delay.delay_len = (u32) AUDIO_SAMPLE_RATE * effect->delay._delay_len_seconds * 2;
+            effect->delay._prev_delay_len_seconds = effect->delay._delay_len_seconds;
+        }
+        for (u32 i = 0; i < len; i++) {
+            //                                                    vvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvvv
+            buffer[(start + i) % CHANNEL_BUFFER_LENGTH] += buffer[ABS(((s32) start + i - effect->delay.delay_len) % CHANNEL_BUFFER_LENGTH)] * effect->delay.feedback;
+        }
+    };
+    Effect effect = {delay_func};
+    effect.delay.feedback = feedback;
+    effect.delay._delay_len_seconds = delay_time;
+    return effect;
+}
+
+Effect* add_effect(Effect effect, u32 channel_id) {
+    ASSERT(channel_id < NUM_CHANNELS, "Invalid channel_id");
+    for (u32 i = 0; i < NUM_EFFECTS; i++) {
+        if (audio_struct.effects[channel_id][i].effect) continue;
+        audio_struct.effects[channel_id][i] = effect;
+        return &audio_struct.effects[channel_id][i];
+    }
+    ERR("Not enough free effect slots on channel %d, skipping effect", channel_id);
+    return nullptr;
+}
+
+void clear_effects(u32 channel_id) {
+    ASSERT(channel_id < NUM_CHANNELS, "Invalid channel_id");
+    for (u32 i = 0; i < NUM_EFFECTS; i++) {
+        audio_struct.effects[channel_id][i].effect = nullptr;
+    }
+}
```

Comment:
This invalidates pointers. We can either use IDs or pass the responsibility of keeping track of pointers to the game developer. I'm leaning towards the latter, since effects are (should be) controlled in the "broader scope" in the game (compared to entities who "update themselves"), but I'm not sure.

Reply:
I'd say give them explicit control of the effect slots. I don't recommend letting people be in charge of memory management since they often do it wrong, knowing where it is and what to test for saves a lot of time. There might be a better solution here somewhere but there isn't a "one solution" for this type of problem. I'd lean more towards IDs.

Reply:
Maybe function pointers are the wrong match for this kind of code, it's possible to write it like the particle system, with a bunch of settings for everything, and bake all of it into one huge struct.

Maybe all effects should just allways "be active" but can be tweaked by the values in the struct.

Or you can make fancy template functions which searches for the type you pass in and returns
a valid pointer if the effect exists. 

Or you add a type field in the Effect struct, and have each effect in a specific slot in a list for each channel. This lets you know exactly where things are.

I'd go for adding types before forcing users of the engine to manage the effects memory. It's usually easier if you have a clear owner of the memory passed around, and currently the engine owns everything, which would make this part of the API stand out.

Reply:
In that case I'd say ID for now (check the latest commits) and see how it looks.

Reply:
I feel like it should be as simple as possible to make your own effects if you want to, which should boil down to something like "write a function that transforms a buffer in some way and attach it to the effects system". We can then export helpers (like an FFT) to make it easier to create additional effects.